feat(spans): Ingest standalone spans

phacops opened the pull request:
This PR aims to add support for ingesting standalone spans. This will support 3 ways to ingest spans:
- an HTTP endpoint accepting OpenTelemetry spans, encoded as JSON
- an `Item` of type `OtelSpan` accepting an OpenTelemetry span, encoded as JSON
- an `Item` of type `Span` accepting a span matching the span protocol, encoded as JSON

After we accept those spans, they are converted to a span as the protocol describes it and they each go through the span processor which will: 
- normalize
- validate
- extract metrics

They will then be sent to Kafka for storage.

In the end, we're aiming for this except our common format is the span protocol:
![image (4)](https://github.com/getsentry/relay/assets/336345/87c653e8-1686-402c-a761-0e81fe968176)

phacops committed:
feat(otel): Add an endpoint to ingest OpenTelemetry spans

phacops committed:
Move process_spans under the processing macro

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

phacops committed:
Ingest a real OTLP span schema

phacops committed:
Handle parent_span_id and segment_id

jjbayer commented on relay-server/src/endpoints/spans.rs:
```diff
@@ -0,0 +1,45 @@
+    let event_span: Annotated<EventSpan> = Annotated::new(span.into());
+
+    // Pack into an envelope for further processing.
+    let span_json = event_span.to_json().map_err(BadStoreRequest::InvalidJson)?;
```
We probably don't want to deserialize or convert anything between request and response, to keep the response time as short as possible (see for example the security report endpoint). Instead, we can add a new item type `OtelSpan` that contains the raw JSON, and then unpack it in `process_` function (as we do for profiles).

jjbayer commented on relay-spans/src/lib.rs:
```diff
@@ -0,0 +1,247 @@
+    pub links: Vec<Link>,
+    pub dropped_links_count: u32,
+    pub status: Option<Status>,
+}
```
For fields that are not required, we should add `#[serde(default)]` and `#[serde(skip_serializing_if)]` attributes.

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2427,6 +2427,29 @@ impl EnvelopeProcessorService {
     }
 
+    #[cfg(feature = "processing")]
+    fn process_spans(&self, state: &mut ProcessEnvelopeState) {
```
I think we also want a `filter_spans` method that drops spans in customer relays and/or PoPs when the feature flag is enabled.

jjbayer commented on relay-spans/src/lib.rs:
```diff
@@ -0,0 +1,247 @@
+    pub dropped_attributes_count: u32,
+}
+
+impl From<Span> for EventSpan {
```
@phacops we briefly discussed this offline: I'm wondering if instead of converting to `EventSpan`, we should convert event spans to `OtelSpan` instead when we extract them from transactions?

I realize that this would slow down development because we'd have to make sure that the sentry consumer supports the OTEL format first, so I am also fine with converting to event spans for now.

phacops replied:
Yes, that's precisely why I started like this. I realize in the long term, we want the reverse but we're long from being ready for this.

I think in the interest of figuring out how an OTel SDK would interact with this, it's still worth doing this and then reversing when we're ready.

jjbayer replied:
Makes sense!

jjbayer reviewed (commented):
Good stuff! Long-term we'll need a rate limiting and dynamic sampling strategy for these spans, but as long as the ingestion's feature-flagged we can ship without it IMO.

phacops committed:
Emit OtelSpan from the HTTP handler and parse them when processing

phacops committed:
Filter spans if the feature is not active

phacops committed:
Parse attributes' values with the right type

phacops committed:
Use default value instead of Option

phacops committed:
Add data and status fields

phacops committed:
Reorganize the crate

jjbayer commented on relay-dynamic-config/src/feature.rs:
```diff
@@ -31,6 +31,9 @@ pub enum Feature {
     SpanMetricsExtractionResource,
+    /// Enable OTel span ingestion.
+    #[serde(rename = "organizations:otel-span-ingestion")]
+    OTelSpanIngestion,
```
nit

https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case

```suggestion
    OtelSpanIngestion,
```

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2427,6 +2428,59 @@ impl EnvelopeProcessorService {
+                let span: relay_spans::Span = match serde_json::from_slice(&item.payload()) {
+                    Ok(span) => span,
+                    Err(e) => {
+                        relay_log::error!("Invalid span: {e}");
```
```suggestion
                        relay_log::error!(error = &e as &dyn Error, "failed to parse OTel span");
```

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2427,6 +2428,59 @@ impl EnvelopeProcessorService {
+                        let payload: Bytes = match span.to_json() {
+                            Ok(payload) => payload.into(),
+                            Err(e) => {
+                                relay_log::error!("Invalid span: {e}");
```
```suggestion
                                relay_log::error!(
                                    error = &e as &dyn Error,
                                    "Failed to serialize OTel span"
                                );
```

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub trace_state: String,
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub parent_span_id: Option<String>,
```
Could not find a formal OTel schema definition, but it seems that `parent_span_id` is an empty string for root spans: https://opentelemetry.io/docs/concepts/signals/traces/#spans

phacops replied:
I intended to remove the `Option` as you suggested before, so I'll fix this. Are you also suggesting to remove the `skip_serializing_if` in case the string is empty so we propagate the empty value in case of a root span?

jjbayer replied:
> Are you also suggesting to remove the skip_serializing_if in case the string is empty so we propagate the empty value in case of a root span?

No, `skip_serializing_if` is fine here IMO.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+    pub name: String,
+    pub kind: SpanKind,
+    pub start_time_unix_nano: i64,
+    pub end_time_unix_nano: i64,
```
What formal schema are we implementing? Is it this one? https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto.  cc @mydea 

Asking because on [this page](https://opentelemetry.io/docs/concepts/signals/traces/#spans) I only see `start_time` and `end_time` as strings, no `_unix_nano` suffix.

phacops replied:
It's indeed this one we're implementing. All SDKs auto-generated those structures and would send us a JSON structure based on this protobuf file.

The other page you link is the more user friendly implementation to build a span in the SDK.

Ideally, I would want us to use the auto-generated version eventually but they seemed to have only files meant to be used with their GPRC/tonic implementations and I wanted serde + JSON. Eventually it'd be good to migrate to something auto-generated from the protobuf though.

mydea replied:
hmm good point, I generated the example based on what opentelemetry JS exports for OTLP 🤔

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+pub struct Link {
+    pub trace_id: String,
+    pub span_id: String,
+    pub trace_state: String,
```
Same question for all the fields on all the sub-structs: Is this field optional? If so, `#[serde(default)]`.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+            "SPAN_KIND_CONSUMER" => Some(Self::Consumer),
+            _ => None,
+        }
+    }
```
Can we remove these methods? Doesn't seem like we need them anywhere.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+            _ => None,
+        }
+    }
+}
```
Same question about removing these.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+    }
+}
+
+impl Span {
```
nit: let's put this block right after `Span`.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+                .unwrap()
+                .get("sentry.environment")
+                .unwrap()
+                .as_str(),
```
nit: You can probably use `get_path!` here.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,350 @@
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Span {
```
Took me a while to realize that we're implementing [this schema](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto), might be worth linking in the doc comments.

jjbayer commented on relay-server/src/endpoints/spans.rs:
```diff
@@ -0,0 +1,45 @@
+
+    let mut item = Item::new(ItemType::OtelSpan);
+    item.set_payload(ContentType::Json, body);
+    let mut envelope = Envelope::from_request(Some(EventId::new()), meta);
```
```suggestion
    let mut envelope = Envelope::from_request(None, meta);
```

jjbayer commented on relay-spans/src/lib.rs:
```diff
@@ -0,0 +1,4 @@
+pub use crate::span::Span;
```
Let's be pedantic about docs.

```suggestion
#![warn(missing_docs)]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/getsentry/relay/master/artwork/relay-icon.png",
    html_favicon_url = "https://raw.githubusercontent.com/getsentry/relay/master/artwork/relay-icon.png"
)]

pub use crate::span::Span;
```

phacops committed:
Add docs auto-generation

Co-authored-by: Joris Bayer <joris.bayer@sentry.io>

phacops committed:
Add a link to the schema we're implementing

phacops committed:
Omit EventId as it's set later on if needed

phacops committed:
Move span related code right after the span struct

phacops committed:
Remove unwanted Option

phacops committed:
Use better error messages

phacops committed:
Rename feature flag

phacops committed:
Remove unneeded structs

phacops committed:
Use macro to access map value

phacops committed:
Add documentation

phacops committed:
Fix links

phacops committed:
Add more documentation and use default where appropriate

phacops committed:
Fix link and remove examples trying to be parsed during tests

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

phacops committed:
Populate the received field

phacops committed:
Convert OTel attributes to Sentry tags

phacops committed:
Add a CHANGELOG entry

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2489,6 +2490,62 @@ impl EnvelopeProcessorService {
+        let mut items: Vec<Item> = Vec::new();
+        state.managed_envelope.retain_items(|item| match item.ty() {
+            ItemType::OtelSpan => {
+                let span: relay_spans::Span = match serde_json::from_slice(&item.payload()) {
```
nit: Could use [`let .. else`](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html) here.

phacops replied:
Does it mean I need to drop the error logging or is there a way to do it with error logging as well?

jjbayer replied:
Oh right, you can't get the error info that way. Ignore me.

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2489,6 +2490,62 @@ impl EnvelopeProcessorService {
+                        relay_log::error!("Invalid span: {e}");
+                        ItemAction::DropSilently
+                    }
+                }
```
nit: We have a mutable reference to `item` here, so instead of pushing to a new vector, we could just write straight into the item.

phacops replied:
I was doing this to avoid looping on new items.

jjbayer commented on relay-server/src/envelope.rs:
```diff
@@ -162,6 +164,7 @@ impl fmt::Display for ItemType {
             Self::ReplayRecording => write!(f, "replay_recording"),
             Self::CheckIn => write!(f, "check_in"),
             Self::Span => write!(f, "span"),
+            Self::OtelSpan => write!(f, "otel_span"),
```
Please also add an entry to `from_str` below. Unfortunately the compiler does not catch this.

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,447 @@
+/// A Span represents a single operation performed by a single component of the system.
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Span {
```
Should we call this type `OtelSpan`? Just to be super clear what it is. Then at a later point, when we invert the pipeline and convert event spans into otel spans, we can rename this type back to `Span`.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,447 @@
+                attribute.key
+            };
+            match attribute.value {
+                AnyValue::Array(_) => todo!(),
```
Should we fix the `todo!()`s here? We should be able to convert to `Object` or `Array`.

phacops replied:
I removed the `todo!()`. I don't know yet how we need to handle the `Kvlist` or `Array` as we don't support that in our Sentry tags.

We should stringify them one way or another and show their values or flatten them. It's not blocking now because we already don't support them so we can make a decision on this later.

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2489,6 +2490,62 @@ impl EnvelopeProcessorService {
+                    }
+                };
+                let event_span: Annotated<Span> = Annotated::new(span.into());
+                match self.validate_span(event_span) {
```
`validate_span` currently relies on certain validations by the `TransactionProcessor`, see https://github.com/getsentry/relay/blob/dda78339e126f9bf6800861efd5ea871aecb80f6/relay-server/src/actors/processor.rs#L2412-L2416. We should replicate this validation somehow (happy to contribute).

phacops committed:
Simplify span handling

phacops committed:
Add missing entry

phacops committed:
Rename Span to OtelSpan

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

phacops committed:
Extract metrics from standalone spans only

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

phacops committed:
Accept EventSpan as well

phacops committed:
Validate timestamps and IDs on the span

phacops committed:
Fix EventSpan support

phacops committed:
Fix import

phacops committed:
Make sure everything is behind the feature flag

phacops committed:
Remove todos on attribute value handling

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2504,6 +2524,110 @@ impl EnvelopeProcessorService {
+        for item in state.managed_envelope.envelope().items() {
+            if item.ty() == &ItemType::OtelSpan {
+                if let Ok(otel_span) =
+                    serde_json::from_slice::<relay_spans::OtelSpan>(&item.payload())
```
Instead of deserializing again here (we also deserialize in `process_spans`), let's have a more linear pipeline inside `process_spans`, i.e.

```
FOR EACH span item
    1. Convert otel to sentry span if needed
    2. validate / normalize
    3. extract metrics
    4. write back into envelope
```

phacops replied:
So, we'll only extract metrics on processing nodes?

Also, this works while we're not dynamically sampling spans, but when we do, we'll need to split this again.

jjbayer replied:
Yes, I would keep everything in processing nodes for now, so we can iterate fast without having to think about outdated external relays or duplicate metrics extraction.

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2504,6 +2524,110 @@ impl EnvelopeProcessorService {
+            _ => return,
+        };
+        let extract_span_metrics = |span: Annotated<Span>| {
+            let span = match self.validate_span(span) {
```
I will check event normalization to see if any more validation and/or normalization is needed.

jjbayer commented on relay-server/src/envelope.rs:
```diff
@@ -114,6 +114,10 @@ pub enum ItemType {
+    /// A standalone OpenTelemetry span.
+    OtelSpan,
+    /// A standalone span.
+    EventSpan,
```
I think we can make this work without an additional item type, and use `Span` for ingesting sentry-style spans directly.

```suggestion
```

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2799,10 +2924,16 @@ impl EnvelopeProcessorService {
+        if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
+            self.extract_span_metrics(state);
+        }
+
```
Let's call `extract_span_metrics` from within `process_spans` (see other comments).

```suggestion
```

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2504,6 +2524,110 @@ impl EnvelopeProcessorService {
+        };
+        let standalone_span_ingestion_enabled = state
+            .project_state
+            .has_feature(Feature::StandaloneSpanIngestion);
```
nit: In theory we don't have to check this feature flag here because it is already done in `filter_spans`.

phacops replied:
Could someone send spans through an old Relay version (not having this filtering logic) and somehow have spans be ingested?

I added this logic to potentially protect against cases like that.

jjbayer replied:
> Could someone send spans through an old Relay version (not having this filtering logic) and somehow have spans be ingested?

Those would still get caught by the `filter_spans` function of the newer Relay, so I think it's safe to remove here.

jjbayer reviewed (changes requested):
I will contribute an integration test to see if spans & metrics make it through to kafka.

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

jjbayer committed:
test: integration

phacops committed:
Revert EventSpan addition

phacops committed:
Streamline deserialization, normalization and metrics extraction

phacops committed:
Fix import

phacops committed:
Do not return if the config is disabled

phacops committed:
Add exclusive_time to spans in ingestion integration tests

phacops committed:
Allow metrics extraction

phacops committed:
Add what the metric is supposed to look like

jjbayer committed:
metrics test

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,451 @@
+impl From<OtelSpan> for EventSpan {
+    fn from(from: OtelSpan) -> Self {
+        let end_timestamp = Utc.timestamp_nanos(from.end_time_unix_nano);
+        let exclusive_time = (from.end_time_unix_nano - from.start_time_unix_nano) as f64 / 1e6f64;
```
Are we running into a fundamental problem with span streaming here? Exclusive time cannot be computed without knowing how much time the child spans took up.

phacops replied:
I don't think it's a fundamental problem, SDKs will have to track this themselves and make it an attribute instead.

It's handled like that here because I needed to make a valid span.

phacops committed:
Read exclusive_time from an attribute

phacops committed:
Merge remote-tracking branch 'origin/master' into pierre/spans-ingest-otel-schema

phacops committed:
Refacor span processing into its own file

phacops committed:
Add exclusive_time attributes to span fixtures

phacops committed:
Set received and is_segment for both OTel and Event spans

phacops committed:
Remove unused import

phacops committed:
Fix test

jjbayer committed:
feat(spans): Normalization for span ingestion (#2777)

Run full normalization before extracting metrics from a standalone span.

---------

Co-authored-by: Pierre Massat <pierre.massat@sentry.io>

jjbayer commented on relay-server/src/actors/processor/span.rs:
```diff
@@ -0,0 +1,418 @@
+    pub transaction_range: std::ops::Range<UnixTimestamp>,
+    /// The maximum allowed size of tag values in bytes. Longer values will be cropped.
+    pub max_tag_value_size: usize,
+}
```
Does not need to be `pub` anymore.
```suggestion
/// Config needed to normalize a standalone span.
#[cfg(feature = "processing")]
#[derive(Clone, Debug)]
struct NormalizeSpanConfig {
    /// The time at which the event was received in this Relay.
    pub received_at: DateTime<Utc>,
    /// Allowed time range for transactions.
    pub transaction_range: std::ops::Range<UnixTimestamp>,
    /// The maximum allowed size of tag values in bytes. Longer values will be cropped.
    pub max_tag_value_size: usize,
}
```

jjbayer commented on relay-server/src/actors/processor/span.rs:
```diff
@@ -0,0 +1,418 @@
+
+/// We do not extract spans with missing fields if those fields are required on the Kafka topic.
+#[cfg(feature = "processing")]
+pub fn validate(mut span: Annotated<Span>) -> Result<Annotated<Span>, anyhow::Error> {
```
Can be private now.

```suggestion
fn validate(mut span: Annotated<Span>) -> Result<Annotated<Span>, anyhow::Error> {
```

jjbayer commented on relay-server/src/actors/processor/span.rs:
```diff
@@ -0,0 +1,418 @@
+}
+
+#[cfg(feature = "processing")]
+pub fn extract_from_event(state: &mut ProcessEnvelopeState) {
```
Let's move all public functions to the top of the file.

jjbayer commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 **Features**:
 
+- Ingest OpenTelemetry spans via HTTP or an envelope. ([#2620](https://github.com/getsentry/relay/pull/2620))
```
```suggestion
- Ingest OpenTelemetry and standalone Sentry spans via HTTP or an envelope. ([#2620](https://github.com/getsentry/relay/pull/2620))
```

jjbayer commented on relay-server/src/actors/processor/span.rs:
```diff
@@ -0,0 +1,418 @@
+    // `is_segment` is set by `extract_span`.
+    exclusive_time
+        .value()
+        .ok_or(anyhow::anyhow!("missing exclusive_time"))?;
```
This check means that we only accept spans for which the SDK sets `exclusive_time`. So only from a Sentry SDK, not any other client.

phacops replied:
I changed it so the fallback is to calculate the duration of the span using start and end timestamps. Like this, the validation still works as intended.

jjbayer replied:
See other comment:

> This is ultimately a product decision, but I don't think we should assume that a span has zero child spans just because the sentry-specific "exclusive time" attribute is missing. I'd rather drop these spans for now than ingest incorrect data for them.

jjbayer commented on relay-spans/src/otel_to_sentry_tags.rs:
```diff
@@ -0,0 +1,36 @@
+        ),
+        ("http.request.method", "request.method"),
+        ("sentry.environment", "environment"),
+        ("sentry.exclusive_time_ns", "exclusive_time_ns"),
```
TODO: Why is there a mapping for this?

phacops replied:
I thought it might be a good idea to have it but thinking about it more, it's useless, so I removed it.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,459 @@
+}
+
+#[derive(Clone, Debug, Deserialize_repr, Default, PartialEq)]
+#[repr(u32)]
```
```suggestion
#[repr(u8)]
```

Any reason for `u32` here?

phacops replied:
I was about to say I wanted to stick with what was [generated](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-proto/src/proto/grpcio/opentelemetry.proto.trace.v1.rs#L305C1-L305C1) but turns out, I'm even changing that to unsigned integer, so might as well use a `u8`.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,459 @@
+    pub code: StatusCode,
+}
+
+#[derive(Clone, Debug, Deserialize_repr, Default, PartialEq)]
```
```suggestion
#[derive(Clone, Copy, Debug, Deserialize_repr, Default, PartialEq)]
```

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,459 @@
+
+        if status_code == StatusCode::Unset || status_code == StatusCode::Ok {
+            return "ok";
+        }
```
Might be a matter or personal taste, but why not

```suggestion
        if self.status.code == StatusCode::Unset || self.status.code == StatusCode::Ok {
            return "ok";
        }
```

phacops replied:
I wanted to avoid the `clone` twice but since you added `Copy` above, I guess it's handled for me so sure.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,459 @@
+            .attributes
+            .clone()
+            .into_iter()
+            .find(|a| a.key == "http.status_code")
```
```suggestion
        if let Some(code) = self.attributes.iter().find(|a| a.key == "http.status_code") {
```

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,459 @@
+            .attributes
+            .clone()
+            .into_iter()
+            .find(|a| a.key == "rpc.grpc.status_code")
```
```suggestion
        if let Some(code) = self
            .attributes
            .iter()
            .find(|a| a.key == "rpc.grpc.status_code")
```

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -0,0 +1,459 @@
+        let mut attributes: Object<Value> = Object::new();
+        let start_timestamp = Utc.timestamp_nanos(from.start_time_unix_nano);
+        let end_timestamp = Utc.timestamp_nanos(from.end_time_unix_nano);
+        for attribute in from.attributes.clone() {
```
To prevent unnecessary clones, deconstruct and then use the fields directly:

```suggestion
        let OtelSpan {
            trace_id,
            span_id,
            parent_span_id,
            name,
            attributes: from_attributes,
            ..
        } = from;

        for attribute in from_attributes.clone() {
```

phacops replied:
I tried to push that as much as I could, I still needed some `clone()` for attributes, let me know if I actually made the situation completely better or just changed it.

jjbayer replied:
👍 I pushed a commit to reduce clones even further.

jjbayer reviewed (commented):
Almost there!

phacops committed:
Update CHANGELOG.md

Co-authored-by: Joris Bayer <joris.bayer@sentry.io>

phacops committed:
Make NormalizeSpanConfig private

Co-authored-by: Joris Bayer <joris.bayer@sentry.io>

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

phacops committed:
Move all public functions to the top of the file

phacops committed:
Less clone, more perform

phacops committed:
Use attributes to get exclusive time or default to duration

phacops committed:
Parse a proper OTel trace on HTTP ingestion

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema

jjbayer committed:
ref: fewer clones

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -112,59 +112,67 @@ pub struct OtelSpan {
             };
-            if key == "exclusive_time_ns" {
-                let value = match attribute.value {
+            if key.contains("exclusive_time_ns") {
```
Is `.contains` really what we want here? Don't we want to check for a specific key (or a set of keys)?

phacops replied:
We can check for 2 specific keys indeed, I thought that'd be fine to just use `contains` instead of repeating nearly exact key names.

jjbayer commented on relay-spans/src/span.rs:
```diff
@@ -173,52 +180,62 @@ impl From<OtelSpan> for EventSpan {
-        let mut span = EventSpan {
-            data: attributes.into(),
-            description: from.name.clone().into(),
+        if exclusive_time_ms == 0f64 {
+            exclusive_time_ms =
+                (from.end_time_unix_nano - from.start_time_unix_nano) as f64 / 1e6f64;
+        }
```
This is ultimately a product decision, but I don't think we should assume that a span has zero child spans just because the sentry-specific "exclusive time" attribute is missing. I'd rather drop these spans for now than ingest incorrect data for them. 

```suggestion
```

phacops replied:
I think I'd rather keep them for now. Obviously, this is something that will change as we understand how to use this better from an OTLP point of view but I wouldn't reject them right now.

This is not meant to be used right away by customers and what we do with this field has to change before we release it but it's important to let us ingest some data, maybe a bit wrong at first, to continue testing.

jjbayer reviewed (approved):
I still have some reservations around the handling of "exclusive time", but other than that, this looks good to me!

phacops committed:
Merge branch 'master' into pierre/spans-ingest-otel-schema
